spinoso-time/src/time/tzrs/offset.rs

```diff
@@ -0,0 +1,222 @@
+use regex::Regex;
+use tz::timezone::{LocalTimeType, TimeZoneRef};
+use tzdb::local_tz;
+use tzdb::time_zone::etc::GMT;
+
+const SECONDS_IN_MINUTE: i32 = 60;
+const SECONDS_IN_HOUR: i32 = SECONDS_IN_MINUTE * 60;
+
+/// tzdb provides [`local_tz`] to get the local system timezone. If this ever fails, we can
+/// assume `GMT`. `GMT` is used instead of `UTC` since it has a [`time_zone_designation`] - which
+/// if it is an empty string, then it is considered to be a UTC time.
+///
+/// Note: this matches MRI Ruby implmentation. Where `TZ="" ruby -e "puts Time::now"` will return a
+/// new _time_ with 0 offset from UTC, but still still report as a non utc time.
+///
+/// [`local_tz`]: https://docs.rs/tzdb/latest/tzdb/fn.local_tz.html
+/// [`time_zone_designation`]: https://docs.rs/tz-rs/0.6.9/tz/timezone/struct.LocalTimeType.html#method.time_zone_designation
+#[inline]
+#[must_use]
+fn local_time_zone() -> TimeZoneRef<'static> {
+    match local_tz() {
+        Some(tz) => tz,
+        None => GMT,
+    }
+}
+
+/// Represents the number of seconds offset from UTC
+#[allow(variant_size_differences)]
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum Offset {
+    /// UTC offset, zero offset, Zulu time
+    Utc,
+    /// Fixed offset from UTC
+    ///
+    /// Note: A fixed offset of 0 is different from UTC time
+    Fixed([LocalTimeType; 1]),
+    /// A time zone based offset
+    Tz(TimeZoneRef<'static>),
+}
+
+impl<'a> Offset {
+    /// Generate a UTC based offset
+    #[inline]
+    #[must_use]
+    pub fn utc() -> Self {
+        Self::Utc
+    }
+
+    /// Generate an offset based on the detected local time zone of the system
+    ///
+    /// Detection is done by [`tzdb::local_tz`], and if it fails will return a GMT timezone
+    ///
+    /// [`tzdb::local_tz`]: https://docs.rs/tzdb/latest/tzdb/fn.local_tz.html
+    #[inline]
+    #[must_use]
+    pub fn local() -> Self {
+        Self::Tz(local_time_zone())
+    }
+
+    /// Generate an offset with a number of seconds from UTC.
+    #[inline]
+    #[must_use]
+    pub fn fixed(offset: i32) -> Self {
+        let sign = if offset.is_negative() { '-' } else { '+' };
+        let hours = offset / 3600;
+        let minutes = (offset / 60) % 60;
+        let offset_name = format!("GMT {}{}{}", sign, hours, minutes);
+        let local_time_type =
+            LocalTimeType::new(offset, false, Some(offset_name.as_bytes())).expect("Couldn't create fixed offset");
+        Self::Fixed([local_time_type])
+    }
```

Comment:
This should be correct, however writing a test for these values is difficult because of the whole borrowing temporary values etc.